paasta_tools/deployd/queue.py

```diff
@@ -76,29 +86,60 @@ def get(self):  # TODO: timeout?
                     if entry is not None:
                         break
 
-                self.local_state_condition.wait()
+                next_upcoming_wait_until = self._get_next_upcoming_wait_until()
+                cond_wait_until = min(
+                    timeout_timestamp,
+                    next_upcoming_wait_until,
+                    time.time() + MAX_SLEEP_TIME,
+                )
+                hit_timeout = not self.local_state_condition.wait(
+                    timeout=cond_wait_until - time.time()
+                )
+                if hit_timeout and time.time() >= timeout_timestamp:
+                    raise Empty()
 
         entry_data, entry_stat = entry
 
         try:
-            yield entry_data
+            yield self._parse_data(entry_data)
         except Exception:
             self._release(first_available_entry_node)
             raise
         else:
             self._consume(first_available_entry_node)
 
+    def _parse_data(self, entry_data: bytes) -> ServiceInstance:
+        return ServiceInstance(**json.loads(entry_data.decode("utf-8")))
+
+    def _parse_entry_node(self, path: str) -> Tuple[float, float]:
+        basename = path.split("/")[-1]
+        _, priority, wait_until, _ = basename.split("-", maxsplit=4)
+        return float(priority), float(wait_until)
+
     def _get_first_available_entry_node(self):
         for entry_node in self.entry_nodes:
             if entry_node not in self.locked_entry_nodes:
-                return entry_node
+                _, wait_until = self._parse_entry_node(entry_node)
+                now = time.time()
+                if wait_until <= now:
+                    return entry_node
         return None
 
-    def _lock_and_get_entry(self, entry_node):
+    def _get_next_upcoming_wait_until(self) -> float:
+        next_upcoming_wait_until = float("inf")
+        for entry_node in self.entry_nodes:
+            if entry_node not in self.locked_entry_nodes:
+                _, wait_until = self._parse_entry_node(entry_node)
+                next_upcoming_wait_until = min(next_upcoming_wait_until, wait_until)
+        return next_upcoming_wait_until
+
+    def _lock_and_get_entry(self, entry_node: str):
         try:
             lock_path = f"{self.locks_path}/{entry_node}"
+            self.locked_entry_nodes.add(entry_node)
             self.client.create(lock_path, value=self.id, ephemeral=True)
         except NodeExistsError:
+            self.locked_entry_nodes.add(entry_node)
             return None
```

Comment:
Do we actually need the 2nd `self.locked_entry_nodes.add(entry_node)`?    Only `self.client.create()` can raise `NodeExistsError`, in which case we'll have already added the node to `self.locked_entry_nodes`.

Also, anyway, it seems like the watcher should run shortly after this in either case.  Either we got the lock in which case the watcher will start and wait to acquire the condition's lock, or we fail to get the lock, which probably means that the watcher is currently waiting to run.  Maybe we don't even need this?

Reply:
I forget which test I was working on when I added this, but I think it's reasonable: if we get a NodeExistsError here, we know that this entry_node is already locked.

And since `_lock_and_get_entry` is run while `self.local_state_condition` is held, I think there shouldn't be any weird race conditions here.